test_storage_dict is flaky (integration tests with nginx are flaky)

https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=76274&sha=7b34cc2e5dd36e1ab3dc7e6d42b56ee7c66ecb66&name_0=PR&name_1=Integration%20tests%20%28aarch64%2C%204%2F4%29

nginx does not start sometimes:

```
            if not nothrow:
>               raise Exception(
                    f"Command [{shell_args}] return non-zero code {res.returncode}: {res.stderr.decode('utf-8')}"
                )
E               Exception: Command [docker compose --project-name rootteststoragedict-gw3 --env-file /ClickHouse/tests/integration/test_storage_dict/_instances-0-gw3/.env --file /ClickHouse/tests/integration/helpers/../../../tests/integration/compose/docker_compose_nginx.yml --verbose up -d --renew-anon-volumes] return non-zero code 1: time="2025-02-17T14:08:07Z" level=trace msg="Docker Desktop integration not enabled"
E                Network rootteststoragedict-gw3_default  Creating
E                Network rootteststoragedict-gw3_default  Created
E                Container rootteststoragedict-gw3-nginx-1  Creating
E                Container rootteststoragedict-gw3-nginx-1  Created
E                Container rootteststoragedict-gw3-nginx-1  Starting
E               time="2025-02-17T14:08:07Z" level=debug msg="otel error" error="<nil>"
E               time="2025-02-17T14:08:07Z" level=debug msg="otel error" error="<nil>"
E               Error response from daemon: driver failed programming external connectivity on endpoint rootteststoragedict-gw3-nginx-1 (ddc643d9717060270efc2232ce9055da70582f3e432555a8c59ac52d137c2b2f): Bind for 0.0.0.0:80 failed: port is already allocated

```

--- a/tests/integration/compose/docker_compose_nginx.yml
+++ b/tests/integration/compose/docker_compose_nginx.yml
@@ -7,4 +7,8 @@ services:
         image: clickhouse/nginx-dav:${DOCKER_NGINX_DAV_TAG:-latest}
         restart: always
         ports:
-            - 80:80
+            # NOTE: It is enough to change only the host port, since in
+            # container there should not be overlaps, plus, you don't have to
+            # change all the invocations inside the container network to this
+            # $NGINX_EXTERNAL_PORT and simply can use "80".
+            - ${NGINX_EXTERNAL_PORT}:80

--- a/tests/integration/helpers/cluster.py
+++ b/tests/integration/helpers/cluster.py
@@ -639,7 +639,7 @@ class ClickHouseCluster:
         # available when with_nginx == True
         self.nginx_host = "nginx"
         self.nginx_ip = None
-        self.nginx_port = 80
+        self._nginx_port = None
         self.nginx_id = self.get_instance_docker_id(self.nginx_host)
 
         # available when with_redis == True
@@ -772,6 +772,13 @@ class ClickHouseCluster:
 
     def compose_cmd(self, *args: str) -> List[str]:
         return ["docker", "compose", "--project-name", self.project_name, *args]
+
+    @property
+    def nginx_port(self):
+        if self._nginx_port:
+            return self._nginx_port
+        self._nginx_port = self.port_pool.get_port()
+        return self._nginx_port
 
     @property
     def kafka_port(self):
@@ -1491,6 +1498,7 @@ class ClickHouseCluster:
     def setup_nginx_cmd(self, instance, env_variables, docker_compose_yml_dir):
         self.with_nginx = True
 
+        env_variables["NGINX_EXTERNAL_PORT"] = str(self.nginx_port)
         self.base_cmd.extend(
             ["--file", p.join(docker_compose_yml_dir, "docker_compose_nginx.yml")]
         )

